### ISSUE TITLE
mca_fcoll_two_phase makefile is not sufficient

### ISSUE THREAD

**Issue by hppritcha** [open-mpi/ompi#4535]
On master if I run ompi_info, I see this near the top of the output:

```
hpp@otono:~/ompi> (master)ompi_info
[otono:17261] mca_base_component_repository_open: unable to open mca_fcoll_two_phase: /home/hpp/ompi/install/lib/openmpi/mca_fcoll_two_phase.so: undefined symbol: mca_io_ompio_overwrite_amode (ignored)
```

The problem is this variable really needs to be in the common ompio lib, and the makefile for
mca_fcoll_two_phase should have a dependency on this common ompio lib.

**Comment by edgargabriel** [comment]
I'll look into this. It worked on my system, but I understand where the problem might be coming from, and there is a very simple solution in that case.

**Comment by edgargabriel** [comment]
I am sorry, I am completely confused. WHat precisely did you and what caused the problems? The structure is basically unchanged for ~ 1year, it is the same in master and 3.0 series. What symbols are undefined in your build?

**Comment by ggouaillardet** [comment]
this is an abstraction violation. `fcoll/two_phase` is using a symbol defined in the `io/ompio` component.
as far as i am concerned, `common/ompio` is not involved in this issue.

**Comment by ggouaillardet** [comment]
here is a possible fix
```patch
From 09bcf1c9da1407573452e1b5824da9d5c1d0641d Mon Sep 17 00:00:00 2001
From: Gilles Gouaillardet <gilles@rist.or.jp>
Date: Mon, 27 Nov 2017 11:53:25 +0900
Subject: [PATCH] fcoll/two_phase: fix an abstraction violation

add the f_overwrite_amode flag to mca_io_ompio_file_t,
and set it to the value of mca_io_ompio_overwrite_amode value
so this can be accessed from fcoll/two_phase.

Signed-off-by: Gilles Gouaillardet <gilles@rist.or.jp>
---
 ompi/mca/fcoll/two_phase/fcoll_two_phase_file_write_all.c | 2 +-
 ompi/mca/io/ompio/io_ompio.h                              | 1 +
 ompi/mca/io/ompio/io_ompio_component.c                    | 3 ++-
 3 files changed, 4 insertions(+), 2 deletions(-)

diff --git a/ompi/mca/fcoll/two_phase/fcoll_two_phase_file_write_all.c b/ompi/mca/fcoll/two_phase/fcoll_two_phase_file_write_all.c
index 8c62099..d76208c 100644
--- a/ompi/mca/fcoll/two_phase/fcoll_two_phase_file_write_all.c
+++ b/ompi/mca/fcoll/two_phase/fcoll_two_phase_file_write_all.c
@@ -1074,7 +1074,7 @@ static int two_phase_exchage_data(mca_io_ompio_file_t *fh,
 		fh->f_io_array[0].memory_address = write_buf;
 		if (fh->f_num_of_io_entries){
                     if ( fh->f_amode & MPI_MODE_WRONLY &&
-                         !mca_io_ompio_overwrite_amode ){
+                         !fh->f_overwrite_amode ){
                         if ( 0 == fh->f_rank ) {
                             printf("\n File not opened in RDWR mode, can not continue."
                                    "\n To resolve this problem, you can either \n"
diff --git a/ompi/mca/io/ompio/io_ompio.h b/ompi/mca/io/ompio/io_ompio.h
index ccb97f5..ad31dc8 100644
--- a/ompi/mca/io/ompio/io_ompio.h
+++ b/ompi/mca/io/ompio/io_ompio.h
@@ -234,6 +234,7 @@ struct mca_io_ompio_file_t {
     enum ompio_fs_type     f_fstype;
     ompi_request_t        *f_split_coll_req;
     bool                   f_split_coll_in_use;
+    bool                   f_overwrite_amode;
     /* Place for selected sharedfp module to hang it's data.
        Note: Neither f_sharedfp nor f_sharedfp_component seemed appropriate for this.
     */
diff --git a/ompi/mca/io/ompio/io_ompio_component.c b/ompi/mca/io/ompio/io_ompio_component.c
index ec05aea..0559ca6 100644
--- a/ompi/mca/io/ompio/io_ompio_component.c
+++ b/ompi/mca/io/ompio/io_ompio_component.c
@@ -13,7 +13,7 @@
  * Copyright (c) 2008-2017 University of Houston. All rights reserved.
  * Copyright (c) 2015      Los Alamos National Security, LLC. All rights
  *                         reserved.
- * Copyright (c) 2015      Research Organization for Information Science
+ * Copyright (c) 2015-2017 Research Organization for Information Science
  *                         and Technology (RIST). All rights reserved.
  * Copyright (c) 2016-2017 IBM Corporation. All rights reserved.
  * $COPYRIGHT$
@@ -337,6 +337,7 @@ file_query(struct ompi_file_t *file,
     if (NULL == data) {
         return NULL;
     }
+    data->ompio_fh.f_overwrite_amode = mca_io_ompio_overwrite_amode;
 
     *private_data = (struct mca_io_base_file_t*) data;
 
-- 
1.8.3.1

```

**Comment by edgargabriel** [comment]
I understand fundamentally the issue of the abstraction violation (although we could have a lengthy discussion about the fact, that the flat component hierarchy is simply suboptimal for what ompio would need at this point), but my question was more about how to reproduce it. On my systems, I did a fresh checkout and everything works (otherwise I would not have committed it), so I would like to know on which system or with which configure flags to trigger the problem.

**Comment by edgargabriel** [comment]
I have btw. a simpler solution in my mind for fixing two_phase component, would prefer not to have to add a new element to the mca_io_ompio_file_t structure for every mca parameter.

**Comment by ggouaillardet** [comment]
@edgargabriel up to you, feel free to make an other fix you are comfortable with.

As long as you do __not__ configure with`--disable-dlopen`, simply running `ompi_info` should be enough to evidence the issue.

**Comment by edgargabriel** [comment]
hm, this is interesting, but also makes sense. ompi_info does in fact report the error, but the component itself works if used from an MPI application. The difference probably being, that the ompio module is always loaded first in an MPI app (and thus the symbol is defined), while that is not the case with ompi_info.

Anyway, I think there is a trivial solution for this, since I realized that just for the verification purposes the mca parameter is actually not required in the two_phase component.  Will commit the fix shortly.

**Comment by edgargabriel** [comment]
Should be fixed with pr #4559
